wikimedia/wikipedia-ios
# Update diff view for forthcoming Watchlist feature (part 2)

**Phabricator:** 
https://phabricator.wikimedia.org/T335579 (partially)

### Notes
This PR pushes the existing diff changes a little bit further from where they were at in #4552. I tried to focus mainly on data or connecting logic here rather than any UI work. From the list in the last PR, these are the things I worked on:

**Partial**
- Showing article image if available in header
    - I added the plumbing during the diff setup methods to fetch the article image URL.
- Update Components framework menu button to support post initialization configuration (to handle our client pattern of lazy view configuration)
    - I tried to do the the bare minimum of [changes](https://github.com/wikimedia/wikipedia-ios-components/pull/12) on this one to get the username to show in the buttons. This was mainly just so I could test the contextual menu routing. It's working but totally understandable if we don't want to treat this quick thing as final.

**Complete**
 - Improving next/previous revisions handling around boundaries of first revisions
    - I think this is done now. I refactored a bit around how the diff style mode is determined. Now the diff container will choose the correct style and disable the correct previous / next buttons when you are on the first and second revisions. I kept the logic of using the `single` revision look while on the first revision. I also have this working for deep links or routing from article links as well.
 - (Not mentioned previously) Add routing from username contextual buttons 
    - These buttons should now route to the user page, user talk page, and user contributions page after tapping.

### Test Steps
1. Confirm routing to user page, user talk page, and user contributions page works
2. Find an article with a short history and scroll down to the beginning. Confirm first revision shows single diff view, with previous button disabled. Confirm second revision shows compare view, with both next and previous buttons enabled.
3. Confirm these deep link properly:

    [https://en.wikipedia.org/wiki/Special:MobileDiff/24601](https://en.wikipedia.org/wiki/Special:MobileDiff/24601)  (2nd+ revision, should go to the compare diff view)
    [https://en.wikipedia.org/wiki/Special:MobileDiff/1150823759...1162159587](https://en.wikipedia.org/wiki/Special:MobileDiff/1150823759...1162159587)  (2nd+ revision, should go to the compare diff view)
    [https://en.wikipedia.org/wiki/Special:MobileDiff/670565873](https://en.wikipedia.org/wiki/Special:MobileDiff/670565873) (first revision, should go to the single diff view)

    I also threw some other links at the end of https://test.wikipedia.org/wiki/IOS. You can load that article up and tap them to confirm routing.

4. Go to any diff screen on an article with a lead image. Confirm you see lead image URL printed out in console during setup.
